sonic_platform_base/sonic_xcvr/fields/xcvr_field.py

```diff
@@ -0,0 +1,209 @@
+    def __init__(self, name, offset, ro):
+        self.name = name
+        self.offset = offset
+        self.ro = ro
+
+    def get_fields(self):
+        """
+        Return: dict containing all fields nested within this field
+        """
+        fields = {}
+        if hasattr(self, "fields"):
+            for field in self.fields:
+                fields[field.name] = field
+                fields.update(field.get_fields())
+        return fields
+
+    def get_offset(self):
+        """
+        Return: absolute byte offset of field in memory map
+        """
+        return self.offset
+
+    def get_size(self):
+        """
+        Return: size of field in bytes (min. 1)
+        """
+        raise NotImplementedError
+
+    def read_before_write(self):
+        """
+        Return: True if a field needs to be read before written to, False otherwise
+        """
+        raise NotImplementedError
+
+    def decode(self, raw_data):
+        """
+        raw_data: bytearray of length equal to size of field
+        Return: decoded data (high level meaning)
+        """
+        raise NotImplementedError
+
+    def encode(self, val, raw_state=None):
+        """
+        val: data with high level meaning
+        raw_state: bytearray denoting the current state of memory corresponding to this field
+        Return: bytearray of length equal to size of field
+
+        Not implemented if not appropriate for the field (e.g. read-only)
+        """
+        raise NotImplementedError
+
+
+class RegBitField(XcvrField):
+    """
+    Field denoting a single bit. Must be defined under a parent RegField
+    """
+    def __init__(self, name, bitpos, parent=None, ro=True, offset=None):
+        super(RegBitField, self).__init__(name, offset, ro)
+        self.bitpos = bitpos
+        self.parent = parent
+
+    def get_size(self):
+        return 1
+
+    def read_before_write(self):
+        return True
+
+    def decode(self, raw_data):
+        return bool((raw_data[0] >> self.bitpos) & 1)
+
+    def encode(self, val, raw_state=None):
+        assert not self.ro and raw_state is not None
+        curr_state = raw_state[0]
+        if val:
+            curr_state |= (1 << self.bitpos)
+        else:
+            curr_state &= ~(1 << self.bitpos)
+        return bytearray([curr_state])
+
+
+class RegField(XcvrField):
+    """
+    Field denoting one or more bytes, but logically interpreted as one unit (e.g. a 4-byte integer)
+    """
+    def __init__(self, name, offset, *fields, **kwargs):
+        super(RegField, self).__init__(name, offset, kwargs.get("ro", True))
+        self.fields = fields
+        self.size = kwargs.get("size", 1)
+        self._updateBitOffsets()
+
+    def _updateBitOffsets(self):
+        for field in self.fields:
+            field.offset = self.offset + field.bitpos // 8
+
+    def get_size(self):
+        return self.size
+
+    def read_before_write(self):
+        return False
+
+class NumberRegField(RegField):
```

Comment:
That section is not relevant for the current platform API and so it would be out of scope for this PR and the refactoring effort as a whole. If BER support is needed in the future then floating point number support can be added then.